### PR TITLE
✨: Add support for `canAccess()` function in permission.

### PR DESCRIPTION
**Pull request by afitzek**
This will allow to write simpler permissions, which point
to access depending on other resources access.

Change-type: minor

Depends-On: https://github.com/balena-io-modules/odata-parser/pull/45
Depends-On: https://github.com/balena-io-modules/odata-to-abstract-sql/pull/48

Signed-off-by: Andreas Fitzek <andreas@balena.io>

### CODE REVIEW

File: src/sbvr-api/permissions.ts [hunk 1]
```diff
@@ -24,7 +27,8 @@ import {
 	PermissionParsingError,
 } from './errors';
 import { AnyObject, ApiKey, HookReq, User } from './sbvr-utils';
-import * as uriParser from './uri-parser';
+//import * as uriParser from './uri-parser';
```
Comment (Page-) [inline src/sbvr-api/permissions.ts]: debug

File: src/sbvr-api/permissions.ts [hunk 2]
```diff
@@ -430,16 +472,109 @@ const generateConstrainedAbstractSql = (
 	const collapsedPermissionFilters = collapsePermissionFilters(
 		permissionFilters,
 	);
+
+	return collapsedPermissionFilters;
+};
+
+const getResourceName = (resourceName: string) =>
```
Comment (Page-) [inline src/sbvr-api/permissions.ts]: There's a `sqlNameToODataName` function for this already

File: src/sbvr-api/permissions.ts [hunk 3]
```diff
@@ -430,16 +472,109 @@ const generateConstrainedAbstractSql = (
...
+			);
+			convertToLamda(collapsedParentPermissionFilters, lId);
+			_.set(property, 'lambda', {
+				method: 'any',
+				identifier: lId,
+				expression: collapsedParentPermissionFilters,
+			});
+
+			compiler.extraBindVars.push(...parentOdata.binds);
+			//odata.binds = _.concat(odata.binds, parentOdata.binds)
```
Comment (Page-) [inline src/sbvr-api/permissions.ts]: debug?

File: src/sbvr-api/permissions.ts [hunk 4]
```diff
@@ -387,12 +391,52 @@ const $checkPermissions = (
 	});
 };
 
-const constrainedPermissionError = new PermissionError();
-const generateConstrainedAbstractSql = (
+const convertToLamda = (filter: AnyObject, identifier: string) => {
+	const replaceOject = (object: AnyObject) => {
+		if (_.isString(object)) {
```
Comment (Page-) [inline src/sbvr-api/permissions.ts]: I think
```suggestion
		if (typeof object === 'string') {
```
is preferable nowadays (typescript validates it from memory)

File: src/sbvr-api/permissions.ts [hunk 5]
```diff
@@ -387,12 +391,52 @@ const $checkPermissions = (
...
+			return;
+		}
+		if (_.isArray(object)) {
+			object.forEach(element => {
+				replaceOject(element);
+			});
+		}
+
+		if (_.has(object, 'name')) {
```
Comment (Page-) [inline src/sbvr-api/permissions.ts]: Are all these `_.has` scattered around actually necessary? It feels like unnecessary lodash usage (lodash is great, it's just it's easy to use even when it's not really necessary)

File: src/sbvr-api/uri-parser.ts [hunk 6]
```diff
@@ -144,7 +144,7 @@ export const memoizedParseOdata = (() => {
 
 const memoizedGetOData2AbstractSQL = memoizeWeak(
 	(abstractSqlModel: _AbstractSQLCompiler.AbstractSqlModel) => {
-		return new OData2AbstractSQL(abstractSqlModel);
+		return new OData2AbstractSQL(abstractSqlModel, {});
```
Comment (Page-) [inline src/sbvr-api/uri-parser.ts]: You can revert this change - the default is already `{}`

File: src/sbvr-api/permissions.ts [hunk 7]
```diff
@@ -387,12 +391,56 @@ const $checkPermissions = (
...
+			delete object.lambda;
+		}
+	};
+
+	replaceObject(filter);
+};
+
+const rewriteSubPermissionBindings = (filter: AnyObject, counter: number) => {
+	const rewrite = (object: AnyObject) => {
+		if (_.isNil(object)) {
```
Comment (Page-) [inline src/sbvr-api/permissions.ts]:
```suggestion
		if (object == null) {
```

File: src/sbvr-api/permissions.ts [hunk 8]
```diff
@@ -387,12 +391,56 @@ const $checkPermissions = (
...
+			return;
+		}
+
+		if (_.isNumber(object.bind)) {
```
Comment (Page-) [inline src/sbvr-api/permissions.ts]:
```suggestion
		if (typeof object.bind === 'number') {
```

File: src/sbvr-api/permissions.ts [hunk 9]
```diff
@@ -387,12 +391,56 @@ const $checkPermissions = (
...
+			object.bind = counter + object.bind;
+		}
+
+		if (_.isArray(object) || _.isObject(object)) {
```
Comment (Page-) [inline src/sbvr-api/permissions.ts]:
```suggestion
		if (Array.isArray(object) || _.isObject(object)) {
```

File: src/sbvr-api/permissions.ts [hunk 10]
```diff
@@ -387,12 +391,56 @@ const $checkPermissions = (
...
 };
 
-const constrainedPermissionError = new PermissionError();
-const generateConstrainedAbstractSql = (
+const convertToLamda = (filter: AnyObject, identifier: string) => {
+	const replaceObject = (object: AnyObject) => {
+		if (typeof object === 'string') {
+			return;
+		}
+		if (_.isArray(object)) {
```
Comment (Page-) [inline src/sbvr-api/permissions.ts]:
```suggestion
		if (Array.isArray(object)) {
```

File: src/sbvr-api/permissions.ts [hunk 11]
```diff
@@ -387,12 +391,56 @@ const $checkPermissions = (
 	});
 };
 
-const constrainedPermissionError = new PermissionError();
-const generateConstrainedAbstractSql = (
+const convertToLamda = (filter: AnyObject, identifier: string) => {
+	const replaceObject = (object: AnyObject) => {
```
Comment (Page-) [inline src/sbvr-api/permissions.ts]: What is this function for?

File: src/sbvr-api/permissions.ts [hunk 12]
```diff
@@ -387,12 +391,56 @@ const $checkPermissions = (
...
+			object.property = _.cloneDeep(object);
+			object.name = identifier;
+			delete object.lambda;
+		}
+	};
+
+	replaceObject(filter);
+};
+
+const rewriteSubPermissionBindings = (filter: AnyObject, counter: number) => {
```
Comment (Page-) [inline src/sbvr-api/permissions.ts]: Isn't this equivalent to `rewriteBinds` in odata-to-abstract-sql?

File: src/sbvr-api/permissions.ts [hunk 13]
```diff
@@ -430,17 +476,116 @@ const generateConstrainedAbstractSql = (
...
+			const targetResource = compiler.NavigateResources(
+				compiler.defaultResource,
+				property.name,
+			);
+
+			const targetResourceName = sqlNameToODataName(
+				targetResource.resource.name,
+			);
+
+			if (_.findIndex(canAccessTrace, v => targetResourceName === v) >= 0) {
```
Comment (Page-) [inline src/sbvr-api/permissions.ts]: What is this for? It seems like it should be a `.includes` to me

File: src/sbvr-api/permissions.ts [hunk 14]
```diff
@@ -430,17 +476,116 @@ const generateConstrainedAbstractSql = (
...
+				parentOdata,
+			);
+
+			if (collapsedParentPermissionFilters === false) {
+				// We reuse a constant permission error here as it will be cached, and
+				// using a single error instance can drastically reduce the memory used
+				throw constrainedPermissionError;
+			}
+
+			const lId = `${lambdaAlias}+${inc}`;
```
Comment (Page-) [inline src/sbvr-api/permissions.ts]: Is `lId` short for `lambdaId`?

File: src/sbvr-api/permissions.ts [hunk 15]
```diff
@@ -430,17 +476,116 @@ const generateConstrainedAbstractSql = (
...
+			inc = inc + 1;
+			rewriteSubPermissionBindings(
+				collapsedParentPermissionFilters,
+				compiler.bindVarsLength + compiler.extraBindVars.length,
+			);
+			convertToLamda(collapsedParentPermissionFilters, lId);
+			_.set(property, 'lambda', {
```
Comment (Page-) [inline src/sbvr-api/permissions.ts]: This can be simplified to
```suggestion
			property.lambda = {
```
right?

File: src/sbvr-api/permissions.ts (hunk 11 above)
Comment (afitzek) [reply src/sbvr-api/permissions.ts]: I'll add a comment to it as well.
Basically we need to inject all occurences of properties for the new lambda function
for example if there is an `or` operator, where different properties of
the target resource are used, we need to change all occurences of these.
We do this in a recursive way to also cover all nested cases like `or [ and, and ]`

File: src/sbvr-api/permissions.ts (hunk 12 above)
Comment (afitzek) [reply src/sbvr-api/permissions.ts]: I think the `rewriteBinds` in `odata-to-abstract-sql` operates on the abstract sql representation, where this operates on the parsed odata representation. So this happens on the odata level before transforming it into abstract sql.

File: src/sbvr-api/permissions.ts [hunk 16]
```diff
@@ -430,17 +481,117 @@ const generateConstrainedAbstractSql = (
...
+			method: 'any',
+			identifier: lambdaId,
+			expression: collapsedParentPermissionFilters,
+		};
+
+		this.extraBindVars.push(...parentOdata.binds);
+
+		canAccessTrace.push(targetResourceName);
+		const value = this.Property(property);
+		_.remove(canAccessTrace, v => targetResourceName === v);
```
Comment (Page-) [inline src/sbvr-api/permissions.ts]: Shouldn't this be just a `canAccessTrace.pop()` as it's a stack?

File: src/sbvr-api/permissions.ts [hunk 17]
```diff
@@ -430,17 +481,117 @@ const generateConstrainedAbstractSql = (
...
+
+		const targetResource = this.NavigateResources(
+			this.defaultResource,
+			property.name,
+		);
+
+		const targetResourceName = sqlNameToODataName(targetResource.resource.name);
+
+		if (_.includes(canAccessTrace, targetResourceName)) {
+			// return a boolen subquery
```
Comment (Page-) [inline src/sbvr-api/permissions.ts]: Why does it need to be a subquery, does returning `['Boolean', true/false]` not work?

File: src/sbvr-api/permissions.ts (hunk 11 above)
Comment (Page-) [reply src/sbvr-api/permissions.ts]: Ah, so this handles adding the lambda alias where previously it didn't exist, that makes sense

File: src/sbvr-api/permissions.ts [hunk 18]
```diff
@@ -387,12 +392,60 @@ const $checkPermissions = (
...
+			return;
+		}
+		if (Array.isArray(object)) {
+			object.forEach(element => {
+				replaceObject(element);
+			});
+		}
+
+		if (object.hasOwnProperty('name')) {
+			object.property = _.cloneDeep(object);
```
Comment (Page-) [inline src/sbvr-api/permissions.ts]: Does this need to be a `cloneDeep`? I think a simple `clone` would work just as well (and be more efficient)

File: package.json [hunk 19]
```diff
@@ -24,8 +24,8 @@
   "dependencies": {
     "@resin/abstract-sql-compiler": "^6.13.1",
     "@resin/lf-to-abstract-sql": "^3.1.1",
-    "@resin/odata-parser": "^1.3.1",
-    "@resin/odata-to-abstract-sql": "^4.3.0",
+    "@resin/odata-parser": "1.4.0",
+    "@resin/odata-to-abstract-sql": "4.4.0-allow-resource-function-registration-058b3b3ec8bf9dd4da68005ec4a4e5caae77f800",
```
Comment (Page-) [inline package.json]: I think this can be updated now

File: src/sbvr-api/permissions.ts (hunk 17 above)
Comment (afitzek) [reply src/sbvr-api/permissions.ts]: No it does not, I have a boolean entry first, but if I remember correctly, it translates a boolean entry into `1` or `0`. If this is used in an `and` statement it becomes in SQL:

```
... AND 1 ...
```
which is not valid SQL.

File: src/sbvr-api/permissions.ts (hunk 18 above)
Comment (afitzek) [reply src/sbvr-api/permissions.ts]: You are probably right. I'll give it a try.

File: src/sbvr-api/permissions.ts (hunk 17 above)
Comment (Page-) [reply src/sbvr-api/permissions.ts]: Ah, well `['Equals', ['Boolean', true], ['Boolean', true]]` should work and still be somewhat preferable

File: src/sbvr-api/permissions.ts [hunk 20]
```diff
@@ -430,17 +481,117 @@ const generateConstrainedAbstractSql = (
...
+		'GET',
+		[],
+		odata.binds.length,
+	);
+
+	let abstractSqlQuery = tree;
+	odata.binds.push(...extraBindVars);
+	const odataBinds = odata.binds;
+
+	abstractSqlQuery = _.clone(abstractSqlQuery!);
```
Comment (Page-) [inline src/sbvr-api/permissions.ts]:
```suggestionS
	odata.binds.push(...extraBindVars);
	const odataBinds = odata.binds;

	const abstractSqlQuery = _.clone(tree!);
```

File: src/sbvr-api/permissions.ts [hunk 21]
```diff
@@ -430,17 +481,117 @@ const generateConstrainedAbstractSql = (
...
+			identifier: lambdaId,
+			expression: collapsedParentPermissionFilters,
+		};
+
+		this.extraBindVars.push(...parentOdata.binds);
+
+		canAccessTrace.push(targetResourceName);
+		const value = this.Property(property);
+		canAccessTrace.pop();
+		return value;
```
Comment (Page-) [inline src/sbvr-api/permissions.ts]: Using a try/finally would make this safer from edge cases/future changes:
```suggestion
		canAccessTrace.push(targetResourceName);
		try {
			return this.Property(property);
		} finally {
			canAccessTrace.pop();
		}
```

File: src/sbvr-api/permissions.ts (hunk 17 above)
Comment (Page-) [reply src/sbvr-api/permissions.ts]: Bump

Comment (afitzek) [reply src/sbvr-api/permissions.ts]: Ah missed that.

File: src/sbvr-api/permissions.ts [hunk 22]
```diff
@@ -430,17 +481,112 @@ const generateConstrainedAbstractSql = (
...
+	const lambdaAlias = randomstring.generate(20);
+	let inc = 0;
+
+	// We need to trace the processed resources, to be able to break
+	// permissions circles.
+	const canAccessTrace: string[] = [resourceName];
+
+	const canAccessFunction: ResourceFunction = function(property: AnyObject) {
+		// remove method property so that we won't loop back here again at this point
+		delete property.method;
```
Comment (Page-) [inline src/sbvr-api/permissions.ts]: If we know we can only be here for a method then we can probably get better typings for `property`

File: src/sbvr-api/permissions.ts [hunk 23]
```diff
@@ -430,17 +481,112 @@ const generateConstrainedAbstractSql = (
...
+		rewriteSubPermissionBindings(
+			collapsedParentPermissionFilters,
+			this.bindVarsLength + this.extraBindVars.length,
+		);
+		convertToLambda(collapsedParentPermissionFilters, lambdaId);
+		property.lambda = {
+			method: 'any',
+			identifier: lambdaId,
+			expression: collapsedParentPermissionFilters,
+		};
```
Comment (Page-) [inline src/sbvr-api/permissions.ts]: Ideally this would be handled by the better typings for property but if not is it possible to add an `as LambdaProperty` or similar type (assuming there's one available? if not then it'll have to stay as-is - I know the odata-parser types are still fairly loose)

File: src/sbvr-api/permissions.ts [hunk 24]
```diff
@@ -430,17 +481,112 @@ const generateConstrainedAbstractSql = (
...
+
+		const targetResource = this.NavigateResources(
+			this.defaultResource,
+			property.name,
+		);
+
+		const targetResourceName = sqlNameToODataName(targetResource.resource.name);
+
+		if (canAccessTrace.includes(targetResourceName)) {
+			// return a boolen subquery
```
Comment (Page-) [inline src/sbvr-api/permissions.ts]:
```suggestion
			// return a boolean subquery that is always false
```
(or `true` and update the returned query if that was the real intention)

File: src/sbvr-api/permissions.ts [hunk 25]
```diff
@@ -387,12 +392,60 @@ const $checkPermissions = (
...
+		if (typeof object === 'string') {
+			return;
+		}
+		if (Array.isArray(object)) {
+			object.forEach(element => {
+				replaceObject(element);
+			});
+		}
+
+		if (object.hasOwnProperty('name')) {
```
Comment (Page-) [inline src/sbvr-api/permissions.ts]: It doesn't make sense to check `object.hasOwnProperty('name')` if `object` is actually an array
```suggestion
		if (Array.isArray(object)) {
			object.forEach(element => {
				replaceObject(element);
			});
		} else if (object.hasOwnProperty('name')) {
```

File: src/sbvr-api/permissions.ts [hunk 26]
```diff
@@ -387,12 +392,60 @@ const $checkPermissions = (
...
 };
 
-const constrainedPermissionError = new PermissionError();
-const generateConstrainedAbstractSql = (
+const convertToLambda = (filter: AnyObject, identifier: string) => {
+	// We need to inject all occurences of properties for the new lambda function
+	// for example if there is an `or` operator, where different properties of
+	// the target resource are used, we need to change all occurences of these.
+	// We do this in a recursive way to also cover all nested cases like or [ and, and ]
+	const replaceObject = (object: AnyObject) => {
```
Comment (Page-) [inline src/sbvr-api/permissions.ts]: It seems like the more accurate typing here would be:
```suggestion
	const replaceObject = (object: string | AnyObject | AnyObject[]) => {
```

File: src/sbvr-api/permissions.ts [hunk 27]
```diff
@@ -387,12 +392,60 @@ const $checkPermissions = (
 	});
 };
 
-const constrainedPermissionError = new PermissionError();
-const generateConstrainedAbstractSql = (
+const convertToLambda = (filter: AnyObject, identifier: string) => {
```
Comment (Page-) [inline src/sbvr-api/permissions.ts]: I think renaming to `lamdaAlias` would make the code a little clearer
```suggestion
const convertToLambda = (filter: AnyObject, lambdaAlias: string) => {
```

File: src/sbvr-api/permissions.ts (hunk 23 above)
Comment (afitzek) [reply src/sbvr-api/permissions.ts]: We don't have such a type yet unfortunately.
